File: new/cpp/15_3sum_hash_map.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // idea: hash map
        // 2 iterations, get two numbers from these two iterations, and see if the target is in 
        // the frequency_counter, where target = -(first number + second number). If it is 
        // in the frequency_counter, then we have a valid triplet. 
        sort(nums.begin(), nums.end());
        unordered_map<int, int> frequency_counter;

        for (int i = 0; i < nums.size(); i++) {
            frequency_counter[nums[i]]++;
        }

        vector<vector<int>> valid_triplet;
        for (int i = 0; i < nums.size(); i++) {
            frequency_counter[nums[i]]--; // because we are currently visiting this value
            if (i > 0 && nums[i] == nums[i - 1]) continue; // because current value
            // and previous value are same, so we can safely assume we have iterated for
            // previous value, if that iteration had any valid triplets it must have been added
            // in other words this prevents duplicates
            for (int j = i + 1; j < nums.size(); j++) {
                frequency_counter[nums[j]]--;
                if (j - 1 > i && nums[j] == nums[j - 1]) continue;
                int target = -(nums[i] + nums[j]);
                if (frequency_counter[target] > 0) {
                    valid_triplet.push_back({nums[i], nums[j], target});
                }
            }

            for (int k = i + 1; k < nums.size(); k++) {
                frequency_counter[nums[k]]++; // restore all the used values from the previous
                // iteration (jth iteration)
            }
        }

        return valid_triplet;
    }
    // time: O(n^2)
    // space: O(m)
};

File: new/python/15_3sum_hash_map.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # idea: hashmap
        # sort the given array
        # create a defaultDict of ints to have as a frequency counter for all the items
        # seen during the current iteration.
        # for the first iteration, we are having the first number of the triplet
        # for the second iteration, we are having the second number of the triplet
        # once we have these two numbers target (3rd number) = -(first num + second num),
        # if the target value is present in the frequency counter, then that is the third
        # number and we have a valid triplet, add it to our result array that has all the valid
        # triplets.
        # also, we can skip the first and second iteration if the previous value is the same
        # as current iteration's value
        # [-1,0,1,2,-1,-4] -> [-4, -1, -1, 0, 1, 2]
        nums.sort()
        frequency_map = defaultdict(int)
        for i in range(len(nums)):
            frequency_map[nums[i]] += 1

        valid_triplets = []
        # first iteration
        for i in range(len(nums)):
            # first thing we should reduce the current iterations number
            # it means we have used this number for the first iteration
            frequency_map[nums[i]] -= 1
            # if current and previous values are same
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            # second iteration
            for j in range(i + 1, len(nums)):
                frequency_map[nums[j]] -= 1
                if j - 1 > i and nums[j] == nums[j - 1]:
                    continue
                target = -(nums[i] + nums[j])
                if frequency_map[target] > 0:  # valid triplet exists
                    valid_triplets.append([nums[i], nums[j], target])

            # once we come out of the second iteration, we are ready to move to the
            # subsequent rounds of first iteration, but we need to update the
            # frequency counter for all the values used in the current second iteration
            for k in range(i + 1, len(nums)):
                frequency_map[nums[k]] += 1

        return valid_triplets
        # time: O(n ^ 2)
        # space: O(m)
